File: perform_string_shifts/string_shift.go
package main

import "strings"

/**
 * Time: O(n)
 * Space: O(n)
 */
func stringShift(s string, shift [][]int) string {
	var res []string
	var count int
	for _, cmd := range shift {
		if cmd[0] == 0 {
			count -= cmd[1]
		} else {
			count += cmd[1]
		}
	}
	count %= len(s)
	offset := 0 - count
	if count > 0 {
		offset = len(s) - count
	}
	res = append(res, s[offset:], s[:offset])
	return strings.Join(res, "")
}

func main() {

}

File: perform_string_shifts/string_shift.py
from typing import List


def string_shift(s: str, shift: List[List[int]]) -> str:
    """
    Time: O(n)
    Space: O(n)
    """
    left_sh = 0
    for direction, amount in shift:
        if direction == 1:
            amount = -amount
        left_sh += amount
    left_sh %= len(s)  # it will change negative to positive
    s = s[left_sh:] + s[:left_sh]
    return s
